Requires Expressions (#58)
cplusplus/SG20: compile-time-programming/requires-expressions.md

# Module name: Requires Expressions

## Overview

<table>
  <thead>
    <th>Level</th>
    <th>Objectives</th>
  </thead>
  <tr>
    <td>Foundational knowledge</td>
    <td>Define and use requires-expressions to check satisfaction of expressions by given parameters</td>
  </tr>
  <tr>
    <td>Advanced</td>
    <td>Define and use requires-expressions to check properties of expressions</td>
  </tr>
  <tr>
    <td>Further studies</td>
    <td></td>
  </tr>
</table>

## Motivation

> Requires-expressions allow a developer to perform compile-time evaluation on the validity of other expressions. These are fundamental to the ability to write concepts.[[Compile-time programming: concepts]][1]

Reviewer: Missing space after `.`; and can you please put newlines every 75-ish columns?

Author: Manual line wrapping makes baby-jesus cry, but for you, anything. :-)

## Topic introduction

Requires-expressions are compile-time predicates which evaluate to true when their specified set of expressions are all valid for a given set of inputs.

## Foundational knowledge: Writing requires-expressions

### Background/Required Knowledge

A student is able to:

* Write and use a function template [[Compile-time programming: function templates]][2]
* Differentiate between expressions and statements

It is helpful if:

> * A student is aware of the constraints imposed by their template definitions

Author: Ahhh, hmmm... was going for something like

be aware that:
```
template<typename T>
void foo(T t) {
    std::cout << t;
}
```
Has requirements that `T` be copyable and that there be some overload matching `operator<<(std::ostream&, T)`

> * A student is aware of the constraints imposed by their template definitions

Author: Is there an alternative wording you can suggest that might help?

> * A student is aware of the constraints imposed by their template definitions

Reviewer: I think it'd be good to avoid the word "constraints" here, as it (now) has a technical meaning. What you're talking about is the informal _structural requirements(?)_ imposed by the bodies of even unconstrained function templates.

I'm not very happy with my phrase "structural requirements." I'm happy with "requirements" over "constraints," but, without an adjective, "requirements" alone is too easy to mistake for something related to `requires`. (A `requires`-clause imposes a constraint, not a requirement: ick.)

> * A student is aware of the constraints imposed by their template definitions

Author: would `A student is aware of the syntactic requirements, both direct and indirect, imposed by their usage of the type within their template.` sound better?

> * A student is aware of the constraints imposed by their template definitions

Reviewer: How about `The student is aware that templates effectively impose syntactic and semantic requirements on their type parameters based on how those parameters are used`; plus `The student is able to give practical examples of syntactic requirements and semantic requirements`?
(Although at some point we slip from "it's nice if the student arrives knowing this" to "this is actually part of the thing we're trying to teach" :))

> * A student is aware of the constraints imposed by their template definitions

Author: I was leaning more into the "It would be nice for the student to know..." where that knowledge creates a simple baseline for understanding the motivation for having such checks. Maybe something more like "The student is aware that attempts to specialize the template with types or values which do not match otherwise unstated assumptions will cause errors within the template... " something like that... need to wake the brain up before I can actually do the proper words...
The point being, perhaps we want to convey the student would benefit from having had compiler errors for trying to use types that dont work with their template definition.

> * A student is aware of the constraints imposed by their template definitions

Reviewer: `The student is aware that attempting to specialize the template with types or values which do not match otherwise unstated assumptions will cause errors within the template.` :+1: I like that.

> * A student is aware of the constraints imposed by their template definitions

Author: change applied

### Student outcomes

A student should be able to:

1. Write a simple-requirement to assert the validity of an expression
2. Write a type-requirement to check the existence of a type by its identifier
3. Write a compound-requirement to test the resulting type of an expression
4. Write a nested-requirement to test the constexpr value of an operation, as opposed to just the semantic validity

> 5. Use a requires-expression within a concept, requires-clause, or `if constexpr` condition

Reviewer 2: I'm not convinced these are foundational.

Author: My intent was to specify the students ability to utilize the requires-expression in _some_ way. These were the only three places I could think of which would supply a motivating usage of a requires expression. Thus, I thought to mention that using the requires expression within at least _one_ of these would be a foundational outcome.

Reviewer: The names such as "compound-requirement" and "nested-requirement" are certainly not foundational, or even needed IMO. Certainly the _instructor_ is not likely to know what these jargon terms mean.  Is it possible to rephrase these without the jargon, e.g.

    1. Write a simple requires-expression that tests whether `t + 1` is well-formed
    2. Write a requires-expression that tests whether the nested type `T::value_type` exists
    3. Write a requires-expression that tests whether the type of `t + 1` is exactly `int`;
        whether it's a scalar type
    4. Write a requires-expression that tests whether `T::value` is `true`
        (not just whether it's well-formed)
    5. Use a requires-expression as a boolean expression within a concept definition,
        requires-clause, or condition

I dunno, am I just trading your preferred jargon ("compound-requirement") for mine ("well-formed")?

Author: I presume our goal is not to _teach_ the instructors, so much as to guide them about what should be taught? I would wholly expect an instructor to need to do some research on the topic in order to teach it, separate from these guides. Through that, shouldnt we use the proper terms so that they can actually stand a chance in learning enough to teach?

### Caveats

> To require expressions which evaluate to a boolean value, such as `sizeof(t) == 4`, to only be satisfied when the result of the expression is `true`, you must use a nested-requirement, such as `requires sizeof(t) == 4;`. Omitting the `requires` results in a simple-requirement, which is satisfied based purely on semantic validity, not on the result of the operation.

Reviewer: Ooh ooh! If we're doing pitfalls, see my talk "Concepts As She Is Spoke" (2018). I've got tons. https://godbolt.org/z/aLyaZb

        template<class T>
        concept False = requires(T t) {
            t.value -> size_t;
        };
        struct S { size_t value; };
        static_assert(!False<S>);

Reviewer: Oh, and again you said `semantic` when you meant `syntactic`.

Author: That's certainly an interesting example. Had to play around with it for a while, before I found a way to express it. I'll add a point-to-cover on the value of adding static_asserts to both positively and negatively test the concept for known types

Author: just realized that's more of a check to be performed when using a concept, which I am trying to not imply as required to use requires-expressions

Reviewer 2: > Ooh ooh! If we're doing pitfalls, see my talk "Concepts As She Is Spoke" (2018). I've got tons. https://godbolt.org/z/aLyaZb
> 
> ```
>     template<class T>
>     concept False = requires(T t) {
>         t.value -> size_t;
>     };
>     struct S { size_t value; };
>     static_assert(!False<S>);
> ```

This won't compile because `t.value -> size_t` isn't valid syntax, so I'm not sure why it's a pitfall?

Reviewer 2: :facepalm: I see why it's a pitfall now. Ignore me.

Author: I'm happy to add the example, but there's a lot to decompress there, and I'm not sure how to succinctly summarize it.

While I'm at it, just how _does_ one properly write a compound-requirement where you would want to remove cv-ref? I can get it via a nested requirement with a decltype and decay_t, but the compound version eludes me...
https://godbolt.org/z/GKrf3j

Reviewer: @rwdougla, re your nerdsnipe: There's no built-in syntax for that. I would do it this way: https://godbolt.org/z/vd8Wdr

    template<class T, class U>
    concept decays_to = std::same_as<std::decay_t<T>, U>;
    ... { t.value } -> decays_to<size_t>; ...
(There's an active C++23 proposal for "built-in _DECAY_COPY_" which would make `{ auto(t.value) } -> std::same_as<size_t> }` also work in this case, but only because our target type `size_t` happens to be copy-constructible.)

Come to think of it, this would also work: https://godbolt.org/z/hKW661

    ... { std::as_const(t.value) } -> std::same_as<const size_t&>; ...
But I _really_ wouldn't recommend that. It requires the programmer to think inside-out from what they're presumably really trying to do. (Which is maybe the real lesson here. Are you sure you _want_ to write such a requirement? Why? What does it actually let you say about the type `T` that will be useful to you in a template definition?)

Reviewer: And since this whole cvref thing seems like maybe an XY problem, let me just rewind and be explicit about the original example, for those who can't find the time to watch the CppCon talk:

    { t.value } -> std::convertible_to<size_t>;  // Well-formed

This says that `decltype((t.value))` ought to be convertible to `size_t`. Awesome. Unsurprising. Just a _soupçon_ of implementation divergence between Clang [wrong] and MSVC/GCC [right] about the proper number of parens in `decltype((t.value))`.

    { t.value } -> size_t;  // Ill-formed, diagnostic required

This says that `decltype((t.value))` ought to satisfy `concept size_t`. It's overwhelmingly likely that no such concept is in scope, and so this requirement would be flagged as a hard error and your program wouldn't ever compile.

    t.value -> size_t;  // Well-formed and SILENTLY WRONG

This says that the syntactically valid expression `t.value->size_t` ought to be well-formed (and we don't care about what type it is). It's overwhelmingly likely that this expression is _not_ well-formed for your particular `T`, so `False<T>` silently evaluates to `false`.

### Points to cover

> * All requires-expression requirements terminate with a semi-colon.

Reviewer: `semicolon`

Author: I believe this was fixed

Reviewer: Oh, this is Github's UI being terrible and showing comments on text that no longer exists in the PR. Okay.

> * simple-requirements are used to check the semantics of an expression.

Reviewer: You mean "syntax," not "semantics."
Personally I would just say "the well-formedness of an expression."

Author: I believe this was fixed

* nested-requirements are introduced with `requires` and primarily used to check the result of an expression computable by the compiler, including concepts or other requires-expressions.
* type-requirements are introduced with `typename` and used to verify the existence of a type with a particular identifier.
* compound-requirements are enclosed in braces and can be used to check the resulting type of an expression.

> * Parameters should be specified without cv-qualifiers, as though passed by-value.

Reviewer: Well, it depends on what you're doing with them. `requires(T& a, const T& b) { { a = b }; }` is perfectly reasonable.
One thing I _would_ mention here is that parameters have the same behavior w.r.t. value categories as if they were function parameters; i.e.,

    requires(T t) { { t = 1 }; }

will be true for `int`, even though `declval<int>() = 1` would be ill-formed. The parameter `t` has a name, so it's an lvalue! This is not surprising if taught right: "all parameter declarations always behave the same and parameter variables are always lvalues."

Author: Removing that suggestion. Can always add back later with augmentation, if desirable.

* Checks are performed by the compiler, not at run time.

> 

Reviewer: I'd add here: The instructor should talk about `requires requires`, and also explain how to avoid ever needing to write it. See https://quuxplusone.github.io/blog/2019/01/15/requires-requires-is-like-noexcept-noexcept/

Author: I believe this was fixed

## Advanced: Advanced requirements

### Background/required knowledge

* All of the above.
* Knowledge of `noexcept`

A student is able to:

* Write a concept [[Compile-time programming: concepts]][1]

### Student outcomes

A student should be able to:

1. Write compound-requirements which test the `noexcept`ness of an expression.
2. Use a concept as the target of a compound-requirement.

### Caveats

### Points to cover

> * Compound-requirements allow the optional ability to test whether an expression is marked as `noexcept`, by using the `noexcept` keyword.

Reviewer: Yikes! I would be much more specific here. This is another of my "Concepts As She Is Spoke" examples. https://godbolt.org/z/p3bWM9

    template<class T>
    concept True = requires(T t) {
        noexcept( t.foo() );
    };
    struct S { int foo() noexcept(false); };
    static_assert(True<S>);

You'll notice a pattern: simple-requirements are the devil and should be avoided at all costs. (But I'm not necessarily saying that that guideline belongs in _this_ document.)

Author: I am not clear what suggestion, if any, is made here. I'd hesitate to fill these documents with all sorts of edge cases, but this may warrant inclusion?

Reviewer: I think just `s/using the noexcept keyword/using a trailing noexcept keyword/` would be sufficient. Arguably the example below is sufficient; but the example is long enough that the reader's eyes may glaze over before getting to the sneaky trailing-noexcept part.

> * If the return-type-requirement of a compound-requirement is a concept, that concept the concept is given the resulting type as the first parameter, followed by the specified parameters in the compound-requirement. `{ ++x } -> C<int>` would substitute `C<decltype((++x)), int>` and check that concept C is satisfied for those parameters.

Reviewer: `s/the concept //`

Author: I believe this was fixed

on the validity of other expressions. These are fundamental to the ability 
to write concepts. [[Compile-time programming: concepts]][1]

## Topic introduction

Requires-expressions are compile-time predicates which evaluate to true 
when their specified set of expressions are all valid for a given set of 
inputs.

## Foundational: Writing requires-expressions

### Background/Required Knowledge

A student is able to:

* Write and use a function template [[Compile-time programming: function templates]][2]
* Differentiate between expressions and statements

It is helpful if:

* A student is aware of the constraints imposed by their template definitions

### Student outcomes

A student should be able to:

1. Write a simple-requirement to assert the validity of an expression
2. Write a type-requirement to check the existence of a type by its identifier
3. Write a compound-requirement to test the resulting type of an expression
4. Write a nested-requirement to test the constexpr value of an operation, as opposed to just the syntactic validity
5. Use a requires-expression within a concept, requires-clause, or `if constexpr` condition

### Caveats

> To require expressions which evaluate to a boolean value, such as 
> `sizeof(t) == 4`, to only be satisfied when the result of the expression is 
> `true`, you must use a nested-requirement, such as 
> `requires sizeof(t) == 4;`. Omitting the `requires` results in a 

Reviewer 3: ```suggestion
To require that one or more expressions, which evaluate to a boolean value 
like `sizeof(t) == 4`, evaluate to `true` a nested-requirement is needed 
(e.g., `requires sizeof(t) == 4;`). Omitting the `requires` results in a 
```

I find the first sentence a bit hard to read/understand. This would be my attempt to reword it.

Reviewer: I would lose the "one or more":
> To require that a particular boolean expression evaluate to `true`, a nested-requirement is needed (e.g., `requires sizeof(t)==4;`). Omitting the `requires` keyword here results in a...

I just filed https://bugs.llvm.org/show_bug.cgi?id=47941 about Clang's lack of diagnostic here, btw.

Reviewer 3: :+1: I adapted the wording.

> simple-requirement, which is satisfied based purely on syntact validity, 

Reviewer: `s/syntact/syntactic/`

* Checks are performed by the compiler, not at run time.
* If covering usage of requires-expression with requires-clause, [[Compile-time programming: requires clause]][3] demonstrate `requires requires` and show how to ever avoid writing it by using a concept. [[Compile-time programming: concepts]][1]

## Main: Advanced requirements

### Background/required knowledge

* All of the above.
* Knowledge of `noexcept`

A student is able to:

* Write a concept [[Compile-time programming: concepts]][1]

### Student outcomes

A student should be able to:

1. Write compound-requirements which test the `noexcept`ness of an expression.
2. Use a concept as the target of a compound-requirement.

### Caveats

### Points to cover

* Compound-requirements allow the optional ability to test whether an expression is marked as `noexcept`, by using the `noexcept` keyword.
  ```
struct S
{
	void foo() noexcept {}
	void bar() {}
};

static_assert(requires(S s) { { s.foo() } noexcept; } ); // Succeeds. s.foo() is noexcept
static_assert(requires(S s) { { s.bar() } noexcept; } ); // Fails. s.bar() is not noexcept
  ```

> * If the return-type-requirement of a compound-requirement is a concept, that concept is given the resulting type as the first parameter, followed by the specified parameters in the compound-requirement. `{ ++x } -> C<int>` would substitute `C<decltype((++x)), int>` and check that concept C is satisfied for those parameters.

Reviewer: (Please manually word-wrap here too.) The double parens on `decltype((...))` threw me, but I now think they're correct; see Clang bug https://bugs.llvm.org/show_bug.cgi?id=45622

# Module name: Requires Expressions

## Overview

<table>
  <thead>
    <th>Level</th>
    <th>Objectives</th>
  </thead>
  <tr>

>     <td>Foundational knowledge</td>

Reviewer 3: ```suggestion
    <td>Foundational</td>
```

    <td>Define and use requires-expressions to check satisfaction of expressions by given parameters</td>
  </tr>
  <tr>

>     <td>Advanced</td>

Reviewer 3: ```suggestion
    <td>Main</td>
```

    <td>Define and use requires-expressions to check properties of expressions</td>
  </tr>
  <tr>

>     <td>Further studies</td>

Reviewer 3: ```suggestion
    <td>Advanced</td>
```

    <td></td>
  </tr>
</table>

## Motivation

Requires-expressions allow a developer to perform compile-time evaluation 
on the validity of other expressions. These are fundamental to the ability 
to write concepts. [[Compile-time programming: concepts]][1]

## Topic introduction

Requires-expressions are compile-time predicates which evaluate to true 
when their specified set of expressions are all valid for a given set of 
inputs.

> ## Foundational knowledge: Writing requires-expressions

Reviewer 3: ```suggestion
## Foundational: Writing requires-expressions
```

`sizeof(t) == 4`, to only be satisfied when the result of the expression is 
`true`, you must use a nested-requirement, such as 
`requires sizeof(t) == 4;`. Omitting the `requires` results in a 
simple-requirement, which is satisfied based purely on syntact validity, 
not on the result of the operation.

### Points to cover

* All requires-expression requirements terminate with a semicolon.
* simple-requirements are used to check that an expression is well-formed.
* nested-requirements are introduced with `requires` and primarily used to check the result of an expression computable by the compiler, including concepts or other requires-expressions.
* type-requirements are introduced with `typename` and used to verify the existence of a type with a particular identifier.
* compound-requirements are enclosed in braces and can be used to check the resulting type of an expression.
* Checks are performed by the compiler, not at run time.
* If covering usage of requires-expression with requires-clause, [[Compile-time programming: requires clause]][3] demonstrate `requires requires` and show how to ever avoid writing it by using a concept. [[Compile-time programming: concepts]][1]

> ## Advanced: Advanced requirements

Reviewer 3: ```suggestion
## Main: Advanced requirements
```

### Background/required knowledge

* All of the above.
* Knowledge of `noexcept`

A student is able to:

* Write a concept [[Compile-time programming: concepts]][1]

### Student outcomes

A student should be able to:

1. Write compound-requirements which test the `noexcept`ness of an expression.
2. Use a concept as the target of a compound-requirement.

### Caveats

### Points to cover

* Compound-requirements allow the optional ability to test whether an expression is marked as `noexcept`, by using the `noexcept` keyword.
  ```
struct S
{
	void foo() noexcept {}
	void bar() {}
};

static_assert(requires(S s) { { s.foo() } noexcept; } ); // Succeeds. s.foo() is noexcept
static_assert(requires(S s) { { s.bar() } noexcept; } ); // Fails. s.bar() is not noexcept
  ```
* If the return-type-requirement of a compound-requirement is a concept, that concept is given the resulting type as the first parameter, followed by the specified parameters in the compound-requirement. `{ ++x } -> C<int>` would substitute `C<decltype((++x)), int>` and check that concept C is satisfied for those parameters.

> ## Further studies

Reviewer 3: ```suggestion
## Advanced
```
